rfcs/NNN-form-associated-mixin.md

```diff
@@ -0,0 +1,377 @@
+---
+Status: Active
+Champions: @justinfagnani
+PR: {{ update_with_pr_number }}
+---
+
+# Form Associated Mixin
+
+A new labs package with helpers for simplifying the creation of form-associated custom elements.
+
+## Objective
+
+Make it easier to build well-behaved and complete form-associated custom
+elements.
+
+### Goals
+
+- Reduced boilerplate needed for an element to participate in an HTML form,
+  including:
+  - Setting the formAssociated flag
+  - Managing the form value and form state
+  - Check and update element validity
+  - Implementing disable, reset, and restore behavior
+- Allow an element to have complete control over its public API.
+- Optionally, allow an element ot opt-in to best-practice public APIs that match
+  built-in form element APIs.
+
+### Non-Goals
+
+- The `FormControl` mixin does not try to emulate the API of a built-in input
+  exactly, because native inputs are designed to allow for consumers to define
+  custom validation. `FormControl` is instead designed for the element author
+  to define validation.
+- This RFC does not aim to cover all form-related element use cases, such as
+  defining a form and managing form state. It it solely concerned with helpers
+  for defining individual form-associated elements, which could be used with
+  form management utilities.
+
+## Motivation
+
+Building a form-associated custom element requires a lot of common boilerplate:
+
+- Setting `static formAssociated = true`
+- Attaching `ElementInternals`
+- Setting a default ARIA role
+- Calling `internals.setFormValue()` when needed
+- Calling `internals.setValidity()` and `internals.checkValidity()` when needed
+- Handling `formDisabledCallback()`
+- Handling `formResetCallback()` and `formStateRestoreCallback()`
+- Requesting a reactive update when state held in internals changes.
+
+This boilplate makes it much harder to implement a form-associated element.
+Element authors should be able to focus on the actually necessary and unique
+parts of their element:
+
+- The element's value and state
+- Validation logic
+
+## Detailed Design
+
+A new `@lit-labs/forms` package will contain several utilities:
+- A `FormAssociated()` mixin that:
+  - Makes an element form associated
+  - Implements the standard form associated custom element callbacks
+  - Provides a validation hook
+  - Adds no public APIs to the element
+- A set of decorators for designating fields as the form value and state
+  - `@formValue()` for the public form value field
+  - `@formState()` for the fields that compose part of the form state
+  - `@formDefaultValue()` for the public form default value field
```

Comment:
Is lib-labs/form handling whether the form is "touched" in some form? Because native `<input>` elements will have their `value` and `checked` change whenever `defaultValue` and `defaultChecked` are changed as long as the input hasn't been _touched_: https://html.spec.whatwg.org/multipage/input.html#the-input-element:concept-fe-dirty-2

If it's a non-goal then maybe be explicit about it (or is it already and I missed it?)